### PR TITLE
Deserialize a response

### PR DESCRIPTION
**Pull request by jkozlowski**
## After this PR
==COMMIT_MSG==

==COMMIT_MSG==

### CODE REVIEW

File: dialogue-annotations-example/src/test/java/com/palantir/myservice/example/MyServiceIntegrationTest.java [hunk 1]
```diff
@@ -148,13 +141,11 @@ public void close() {}
     }
 
     @Test
-    @Disabled
     public void testCustomResponse200() {
         testCustomResponse(200);
     }
 
     @Test
```
Comment (jkozlowski) [inline dialogue-annotations-example/src/test/java/com/palantir/myservice/example/MyServiceIntegrationTest.java:148]: Remaining 2 tests use an endpoint that has a custom param type, which is not supported right now.

File: dialogue-annotations-example/src/test/java/com/palantir/myservice/example/MyServiceIntegrationTest.java [hunk 2]
```diff
@@ -54,20 +53,17 @@
...
-                            .build();
-                    ServicesConfigBlock scb = ServicesConfigBlock.builder()
-                            .defaultSecurity(TestConfigurations.SSL_CONFIG)
-                            .putServices("myServiceDialogue", partialServiceConfiguration)
-                            .build();
-                    DialogueClients.ReloadingFactory factory =
-                            DialogueClients.create(Refreshable.create(scb)).withUserAgent(TestConfigurations.AGENT);
-                    myServiceDialogue = factory.get(MyService.class, "myServiceDialogue");
-                })
-                .isExactlyInstanceOf(UnsupportedOperationException.class);
```
Comment (carterkozak) [inline dialogue-annotations-example/src/test/java/com/palantir/myservice/example/MyServiceIntegrationTest.java:70]: It's happening!

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ArgumentTypesResolver.java [hunk 3]
```diff
@@ -126,38 +101,24 @@ private String planSerDeMethodName(TypeName in) {
...
     }
 
     private Optional<OptionalType> getOptionalType(Element paramContext, TypeMirror typeName) {
-        // "They tryin'a make me use a visitor, but I say: NO, NO NO!"
-        if (!(typeName instanceof DeclaredType)) {
-            return Optional.empty();
-        }
-        DeclaredType declaredType = (DeclaredType) typeName;
-        if (types.isSameType(declaredType, optionalIntType)) {
+        if (context.isSameTypes(typeName, OptionalInt.class)) {
```
Comment (carterkozak) [inline dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ArgumentTypesResolver.java:108]: OptionalDouble + OptionalLong?

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ReturnTypesResolver.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public ReturnTypesResolver(ResolverContext context) {
+        this.context = context;
+    }
+
+    public Optional<ReturnType> getReturnType(
+            EndpointName endpointName, ExecutableElement element, AnnotationReflector requestAnnotation) {
+        TypeMirror returnType = element.getReturnType();
+
+        Optional<TypeMirror> maybeListenableFutureInnerType = getListenableFutureInnerType(returnType);
+        // TODO(12345): Validate deserializer types match
```
Comment (carterkozak) [inline dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ReturnTypesResolver.java:39]: The compiler will also validate this, though perhaps not with quite as helpful a message. Still, beats runtime/reflection.

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java [hunk 5]
```diff
@@ -99,9 +103,21 @@ private MethodSpec clientImpl(EndpointDefinition def) {
...
         def.arguments().forEach(arg -> generateParam(arg).ifPresent(methodBuilder::addCode));
 
-        methodBuilder.returns(def.returns());
+        methodBuilder.returns(def.returns().returnType());
+
+        boolean isAsync = def.returns().asyncInnerType().isPresent();
 
-        methodBuilder.addCode("throw new $T();", UnsupportedOperationException.class);
+        String executeCode = isAsync
+                ? "$L.clients().call($L, $L.build" + "(), $L);"
```
Comment (carterkozak) [inline dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java]:
```suggestion
                ? "$L.clients().call($L, $L.build(), $L);"
```

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java [hunk 6]
```diff
@@ -99,9 +103,21 @@ private MethodSpec clientImpl(EndpointDefinition def) {
...
+                : "$L.clients().callBlocking($L, $L.build(), $L);";
+        CodeBlock execute = CodeBlock.of(
+                executeCode,
+                serviceDefinition.conjureRuntimeArgName(),
+                def.channelFieldName(),
+                REQUEST,
+                def.returns().deserializerFieldName());
+        methodBuilder.addCode(isAsync ? "return " : (def.returns().isVoid() ? "" : "return "));
+        methodBuilder.addCode(execute);
```
Comment (carterkozak) [inline dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java]: Might be easier to read along these lines (untested, might be misremembering the api a bit):

```suggestion
        methodBuilder.addCode(def.returns().isVoid() && !isAsync ? "$L" : "return $L", execute);
```

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ArgumentTypesResolver.java (hunk 3 above)
Comment (jkozlowski) [reply dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ArgumentTypesResolver.java:108]: I'll add those in a separate PR, the serializer and header only works for these 2 as well...

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ReturnTypesResolver.java (hunk 4 above)
Comment (jkozlowski) [reply dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/data/ReturnTypesResolver.java:39]: Ya, but only to the level of the class ipmlementing DeserializerFactory: it won't peel off the types and make sure the return type matches (and especially not in the ListenableFuture case). I started adding this in this PR, but there's some really annoying stuff when dealing with the annotation processor model, where you need to correctly switch between primitives, declared types and stuff. So it turned out more involved that I'd lke for this PR, so deferrring.

**Review by carterkozak — APPROVED** [review]
lgtm. I added a few nit comments, but the change should be fine to merge when you're ready.

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java (hunk 5 above)
Comment (jkozlowski) [reply dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java]: Dang it, why does this not get autoformatted

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java (hunk 6 above)
Comment (jkozlowski) [reply dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java]: ```methodBuilder.addCode(!def.returns().isVoid() || isAsync ? "return $L" : "$L", execute);```

works

**Pull request merged**